### PR TITLE
Shard Redis.

### PR DESCRIPTION
**Pull request by stephanie-wang**
Continued from #509.

### CODE REVIEW

File: src/common/net.h [hunk 1]
```diff
@@ -1,6 +1,9 @@
 #ifndef NET_H
 #define NET_H
 
+#include <string>
+#include <vector>
```
Comment (robertnishihara) [inline src/common/net.h]: These includes are no longer necessary, right?

File: src/common/state/db.h [hunk 2]
```diff
@@ -4,13 +4,18 @@
 #include "common.h"
 #include "event_loop.h"
 
+#include <vector>
```
Comment (robertnishihara) [inline src/common/state/db.h]: Do we need this include?

File: src/common/net.h (hunk 1 above)
Comment (robertnishihara) [inline src/common/net.h]: Or maybe they need to be included in a different file (like `redis.cc`)?

File: src/common/state/redis.cc [hunk 3]
```diff
@@ -633,31 +751,33 @@ void redis_object_table_request_notifications(
...
-  int status = redisAsyncCommandArgv(
-      db->context, redis_object_table_request_notifications_callback,
-      (void *) callback_data->timer_id, num_args, argv, argvlen);
-  free(argv);
-  free(argvlen);
-
-  if ((status == REDIS_ERR) || db->context->err) {
-    LOG_REDIS_DEBUG(db->context,
-                    "error in redis_object_table_subscribe_to_notifications");
+    redisAsyncContext *context = get_redis_context(db, object_ids[i]);
```
Comment (robertnishihara) [inline src/common/state/redis.cc]: Why this change? This is turning a single redis command into potentially a million separate redis commands. Couldn't that be really expensive?

File: src/common/state/redis.cc [hunk 4]
```diff
@@ -779,6 +903,15 @@ void redis_task_table_test_and_update_callback(redisAsyncContext *c,
   redisReply *reply = (redisReply *) r;
   /* Parse the task from the reply. */
   Task *task = parse_and_construct_task_from_redis_reply(reply);
+  if (task == NULL) {
```
Comment (robertnishihara) [inline src/common/state/redis.cc]: What does this change do, it seems like it changes the behavior of the system? Wouldn't it make sense to still call the callback and let the caller decide how to handle this case in the callback?

Also, if this is only possible with actor tasks, then we should assert that the task is an actor task.

File: src/common/state/redis.cc (hunk 3 above)
Comment (robertnishihara) [inline src/common/state/redis.cc]: Oh, I realize now that you're doing this because the different requests need to go to different redis shards. However, in the case where we are requesting 1 million objects, this may be too expensive..

File: src/global_scheduler/global_scheduler.cc [hunk 5]
```diff
@@ -458,15 +458,15 @@ void start_server(const char *node_ip_address,
...
   signal(SIGTERM, signal_handler);
-  /* IP address and port of redis. */
-  char *redis_addr_port = NULL;
+  /* IP address and port of the primary redis instance. */
+  char *redis_primary_addr_port = NULL;
   /* The IP address of the node that this global scheduler is running on. */
   char *node_ip_address = NULL;
   int c;
-  while ((c = getopt(argc, argv, "h:r:")) != -1) {
+  while ((c = getopt(argc, argv, "h:r:t:")) != -1) {
```
Comment (robertnishihara) [inline src/global_scheduler/global_scheduler.cc]: The `-t` is unused, right?

File: src/local_scheduler/local_scheduler.cc [hunk 6]
```diff
@@ -1193,13 +1192,13 @@ int main(int argc, char *argv[]) {
   char *num_workers_str = NULL;
   int c;
   bool global_scheduler_exists = true;
-  while ((c = getopt(argc, argv, "s:r:p:m:ga:h:c:w:n:")) != -1) {
+  while ((c = getopt(argc, argv, "s:r:t:p:m:ga:h:c:w:n:")) != -1) {
```
Comment (robertnishihara) [inline src/local_scheduler/local_scheduler.cc]: The `t` is unused.

File: src/plasma/plasma_manager.cc [hunk 7]
```diff
@@ -1664,10 +1665,10 @@ int main(int argc, char *argv[]) {
   char *master_addr = NULL;
   /* Port number the manager should use. */
   int port = -1;
-  /* IP address and port of state database. */
-  char *db_host = NULL;
+  /* IP address and port of the primary redis instance. */
+  char *redis_primary_addr_port = NULL;
   int c;
-  while ((c = getopt(argc, argv, "s:m:h:p:r:")) != -1) {
+  while ((c = getopt(argc, argv, "s:m:h:p:r:t:")) != -1) {
```
Comment (robertnishihara) [inline src/plasma/plasma_manager.cc]: The `t` is unused.

File: test/runtest.py [hunk 8]
```diff
@@ -1538,7 +1538,8 @@ def testGlobalStateAPI(self):
     task_table = ray.global_state.task_table()
```
Comment (robertnishihara) [inline test/runtest.py]: Let's add another test in `runtest.py` that tests calling `ray.init` with `num_redis_shards=10` or `20` or some large number and then runs some remote functions and gets the results. I don't think we're testing the changes to the `ray.init` code path anywhere.

**Review by robertnishihara — COMMENTED** [review]
This looks great! I have a few comments/questions!

**Comment by robertnishihara** [comment]
I'm looking into the jenkins failure now.

File: src/common/state/redis.cc (hunk 3 above)
Comment (pcmoritz) [inline src/common/state/redis.cc]: Yeah, it is possible to batch everything that goes into a single shard together. Let's do that as an optimization later.

**Comment by robertnishihara** [comment]
Ok, I observed the following crash when running the `many_driver_test` jenkins test.

STDOUT

```
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler[0x40f879]
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z41redis_task_table_test_and_update_callbackP17redisAsyncContextPvS1_+0x88)[0x42b658]
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(redisProcessCallbacks+0x96)[0x44d9c6]
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(aeProcessEvents+0x130)[0x434c20]
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(aeMain+0x2b)[0x43503b]
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(main+0x2f4)[0x40e924]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f783c860830]
/opt/conda/lib/python2.7/site-packages/ray-0.0.1-py2.7.egg/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_start+0x29)[0x40ec49]
```

STDERR

```
[FATAL] (/ray/src/local_scheduler/local_scheduler.cc:565: errno: Operation now in progress) Check failure: ActorID_equal(TaskSpec_actor_id(spec), NIL_ID)
```

**Comment by robertnishihara** [comment]
Also, the most recent jenkins run failed with https://amplab.cs.berkeley.edu/jenkins/job/Ray-PRB/745/console. I've never seen this error before on jenkins and the jenkins don't seem to be flaky, so it's possible there is a problem.

File: src/common/state/redis.cc (hunk 4 above)
Comment (stephanie-wang) [inline src/common/state/redis.cc]: If the task entry is not found in the table, the callback is not called. Previously, it would crash on the next line since `reply->str` was `NULL`.

It should only happen for actor tasks at the moment, but the problem is that we can't tell whether it's an actor task or not without the task entry.

**Comment by robertnishihara** [comment]
This PR will break the web UI. That's completely fine because we need to rewrite the web UI anyway.

File: src/common/redis_module/ray_redis_module.cc [hunk 9]
```diff
@@ -937,6 +941,17 @@ int TaskTableWrite(RedisModuleCtx *ctx,
...
+
+    if (num_clients == 0) {
+      LOG_WARN(
+          "No subscribers received this publish. This most likely means that "
+          "either the intended recipient has not subscribed yet or that the "
+          "pubsub connection to the intended recipient has been broken.");
+      /* This reply will be received by redis_task_table_update_callback or
+       * redis_task_table_add_task_callback in redis.cc, which will then reissue
+       * the command. */
+      RedisModule_ReplyWithError(ctx, "No subscribers received message.");
```
Comment (robertnishihara) [inline src/common/redis_module/ray_redis_module.cc]: This is a bug, forgot a return statement here.

**Comment by shaneknapp** [comment]
test this please
